File: lintcode/python/1492_koko_eating_bananas.py
#!/usr/bin/python -t

# binary search


import sys, math

class Solution:
    """
    @param piles: an array
    @param H: an integer
    @return: the minimum integer K
    """
    def minEatingSpeed(self, piles, H):
        # Write your code here
        m = len(piles)
        start = 1
        end = max(piles)
        #print(start, end)

        while start + 1 < end:
            mid = (start+end)//2
            hour = self.getHour(piles, mid)
            if hour <= H:
                end = mid
            else:
                start = mid

        if self.getHour(piles, start) <= H:
            return start

        return end

    def getHour(self, piles, val):
        hour = 0
        for i in range(len(piles)):
            hour += math.ceil(piles[i]/val)

        return hour


# binary search solution

class Solution:
    """
    @param piles: an array
    @param H: an integer
    @return: the minimum integer K
    """
    def minEatingSpeed(self, piles, H):
        # Write your code here
        n = len(piles)
        r = max(piles)
        l = 1
        
        while l < r:
            mid = l + (r-l)/2
            total = 0
            for p in piles:
                total += (p+mid-1)/mid
            if total > H:
                l = mid+1
            else:
                r = mid
                
        return l

File: lintcode/golang/1492_koko_eating_bananas.go
//NOT AC, but should work.

/**
 * @param piles: an array
 * @param H: an integer
 * @return: the minimum integer K
 */
 
import "sort"
//import "fmt"

func minEatingSpeed (piles []int, H int) int {
    // Write your code here
    sort.Ints(piles)
    l := piles[0]
    r := piles[len(piles)-1]
    
    //fmt.Println(piles)
    for l+1 < r {
        mid := (l+r)/2
        val := getCnt(piles, mid)
        //fmt.Println(mid, val)
        if val <= H {
            r = mid
        } else {
            l = mid
        }
    }
    
    if getCnt(piles, l) <= H {
        return l
    }
    return r
}

func getCnt(piles []int, tartget int) int {
    var ret int
    
    for _, v := range piles {
        if v <= tartget {
            ret += 1
        } else {
            val := v/tartget
            if v%tartget != 0 {
                ret += 1
            }
            ret += val
        }
    }
    
    return ret
}
